lib/storage/batch.go

```diff
@@ -0,0 +1,130 @@
+package storage
+
+import (
+	"sync"
+
+	"github.com/syndtr/goleveldb/leveldb"
+	leveldbIterator "github.com/syndtr/goleveldb/leveldb/iterator"
+	leveldbOpt "github.com/syndtr/goleveldb/leveldb/opt"
+	leveldbUtil "github.com/syndtr/goleveldb/leveldb/util"
+)
+
+type BatchCore struct {
+	sync.RWMutex
+
+	core  LevelDBCore
+	batch *leveldb.Batch
+
+	inserted map[string][]byte
+}
+
+func NewBatchCore(core LevelDBCore) *BatchCore {
+	return &BatchCore{
+		core:     core,
+		batch:    &leveldb.Batch{},
+		inserted: map[string][]byte{},
+	}
+}
+
+func (bb *BatchCore) convertKey(key []byte) string {
+	return string(key)
+}
+
+func (bb *BatchCore) Has(key []byte, opt *leveldbOpt.ReadOptions) (bool, error) {
+	bb.RLock()
+	defer bb.RUnlock()
+
+	var found bool
+	if _, found = bb.inserted[bb.convertKey(key)]; found {
+		return true, nil
+	}
+
+	return bb.core.Has(key, opt)
+}
+
+func (bb *BatchCore) Get(key []byte, opt *leveldbOpt.ReadOptions) (b []byte, err error) {
+	bb.RLock()
+	defer bb.RUnlock()
+
+	var found bool
+	if b, found = bb.inserted[bb.convertKey(key)]; found {
+		return
+	}
+
+	return bb.core.Get(key, opt)
+}
+
+// NewIterator does not work with `BatchBackend`
+func (bb *BatchCore) NewIterator(r *leveldbUtil.Range, opt *leveldbOpt.ReadOptions) leveldbIterator.Iterator {
+	return bb.core.NewIterator(r, opt)
+}
+
+func (bb *BatchCore) Put(key []byte, v []byte, opt *leveldbOpt.WriteOptions) error {
+	bb.Lock()
+	defer bb.Unlock()
+
+	bb.inserted[bb.convertKey(key)] = v
```

Comment:
I think that just put to inserted not `bb.batch.Put` and then `bb.Write`  do loop with `inserted` to `bb.batch.Put` and then batch writing. It can be made more faster and less usage memory in some case. 
(this case is  many put with a key)

Reply:
@anarcher I also consider your approach, but `inserted` is `map` and 'leveldb.Batch' tries to keep the `Put`ting order(https://github.com/syndtr/goleveldb/blob/master/leveldb/batch.go#L88). I think it has reason. Though making `inserted` to be `[]interface{}`, `inserted` is used for checking and getting the `Put`ted content. So, this is the reason to drop the delaying `Put`.

Reply:
I don’t know the reason of batch index with application level. For replay batch, the batch index can be used. It means that if we can use replay batch. 
But l have no idea about the reason. Would you like to explain the reason? Thanks!

Reply:
@anarcher I don't know exactly, but `leveldb.Batch` handles `Put`ed contents in slices. But one thing I think is this kind of case,

1. put A
3. delete A

The 'delete' action must be after 'put'.

Reply:
> @anarcher I don't know exactly, but leveldb.Batch handles Puted contents in slices. But one thing I think is this kind of case,
> 1.  put A
> 2. delete A
> The 'delete' action must be after 'put'.

So, Last db status is `delete A` in this transaction. Is it right? 

`  Start a transaction     ->  put(A) -> put(A) -> delete(A)  ->  End a transaction (Flush/Write(batch) to LevelDB)`

Do we need to make  batch items like `put(A) (1)` and `put(A) (2)`  if we can get intermediate status of A from buffer (`inserted`) in a transaction ?

Reply:
Yes, correct.

Reply:
Why do we need to make batch items of intermediate status of A
I'm not sure but we just make a batch.Delete(A) only,Is it not right?

Reply:
This BatchCore also must act like LevelDBCore, Get returns the latest item. inserted is for this reason.

Reply:
At this time, we only consider the leveldb, so we don’t need too much abstraction, but we will face to leave leveldb to another. At that time, we will apply your interfaces, I think.